### PR TITLE
DFR-4589-Update-Contact-Details-Automatic-Assigment

### PR DESCRIPTION
**Pull request by dawudgovuk**
### Jira link
https://tools.hmcts.net/jira/browse/DFR-4589

### Change description
CT: Update Contact Details Automatically Assign Applicant Solicitor

### Change description
- requires : 
- #2874 
- #2927

### Testing done
Visual testing

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessService.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    private boolean hasApplicantSolicitorChanged(FinremCaseData caseData, FinremCaseData caseDataBefore) {
+        String currentEmail = caseData.getAppSolicitorEmail();
+        String beforeEmail = caseDataBefore.getAppSolicitorEmail();
+        Organisation currentOrg = Optional.ofNullable(caseData.getApplicantOrganisationPolicy())
+            .map(OrganisationPolicy::getOrganisation)
+            .orElse(null);
+        Organisation beforeOrg = Optional.ofNullable(caseDataBefore.getApplicantOrganisationPolicy())
+            .map(OrganisationPolicy::getOrganisation)
+            .orElse(null);
+        boolean isSameOrganisation = Organisation.isSameOrganisation(currentOrg, beforeOrg);
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessService.java]: Simplify it by using `OrganisationPolicy.isSameOrganisation`

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessService.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            }
+            if (caseDataBefore.isApplicantRepresentedByASolicitor()) {
+                assignPartiesAccessService.revokeApplicantSolicitor(caseDataBefore);
+            }
+        }
+    }
+
+    private void updateRespondentSolicitor(FinremCaseData caseData, FinremCaseData caseDataBefore)
+        throws UserNotFoundInOrganisationApiException {
+        if (hasRespondentSolicitorChanged(caseData, caseDataBefore)) {
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessService.java]: same as above - https://github.com/hmcts/finrem-case-orchestration-service/pull/2843/changes#r2984933840

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessService.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            if (caseData.isRespondentRepresentedByASolicitor()) {
+                assignPartiesAccessService.grantRespondentSolicitor(caseData);
+            }
+            if (caseDataBefore.isRespondentRepresentedByASolicitor()) {
+                assignPartiesAccessService.revokeRespondentSolicitor(caseDataBefore);
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessService.java]: new method is missing in this PR

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessService.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        String currentEmail = caseData.getAppSolicitorEmail();
+        String beforeEmail = caseDataBefore.getAppSolicitorEmail();
+        Organisation currentOrg = Optional.ofNullable(caseData.getApplicantOrganisationPolicy())
+            .map(OrganisationPolicy::getOrganisation)
+            .orElse(null);
+        Organisation beforeOrg = Optional.ofNullable(caseDataBefore.getApplicantOrganisationPolicy())
+            .map(OrganisationPolicy::getOrganisation)
+            .orElse(null);
+        boolean isSameOrganisation = Organisation.isSameOrganisation(currentOrg, beforeOrg);
+        return currentEmail != null && !currentEmail.equals(beforeEmail) || !isSameOrganisation;
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessService.java]: boolean emailChanged = !StringUtils.equalsIgnoreCase(beforeEmail, currentEmail);

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessService.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        }
+
+        // Respondent solicitor access update
+        if (hasRespondentSolicitorChanged(caseData, caseDataBefore)) {
+            try {
+                updateRespondentSolicitor(caseData, caseDataBefore);
+            } catch (UserNotFoundInOrganisationApiException e) {
+                String respSolEmail = caseData.getRespondentSolicitorEmailIfRepresented();
+                log.info("There was a problem updating access to respondent solicitor: %s".formatted(respSolEmail));
+                throw new UserNotFoundInOrganisationApiException();
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessService.java]: same as above

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+        // previous applicant was not represented, now is represented
+        ContactDetailsWrapper contactDetailsWrapper = ContactDetailsWrapper.builder()
+            .applicantRepresented(YesOrNo.YES)
+            .applicantSolicitorEmail("new@email.com")
+            .contestedRespondentRepresented(YesOrNo.NO)
+            .build();
+        ContactDetailsWrapper contactDetailsWrapperBefore = ContactDetailsWrapper.builder()
+            .applicantRepresented(YesOrNo.NO)
+            .applicantSolicitorEmail(null)
+            .contestedRespondentRepresented(YesOrNo.NO)
```
[the remark this answers is not in the document]
Comment (so99y) [reply src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessServiceTest.java]: Probably you can mock or spy on method ContactDetailsWrapper.isRespondentRepresentedByASolicitor so you don't have to split the test into contested and consented.

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+
+    @BeforeEach
+    void setUp() {
+        MockitoAnnotations.openMocks(this);
+    }
+
+    static Stream<org.junit.jupiter.params.provider.Arguments> applicantSolicitorScenarios() {
+        return Stream.of(
+            Arguments.of("new@email.com", "org1", "old@email.com", "org2", true),
+            Arguments.of("same@email.com", "org1", "same@email.com", "org1", false)
```
[the remark this answers is not in the document]
Comment (so99y) [reply src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessServiceTest.java]: and also test on emails with null and empty string

**Review by so99y — CHANGES_REQUESTED** [review]

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+        FinremCallbackRequest finremCallbackRequest = buildCallbackRequestApplicantSolicitor(currentEmail, currentOrgId,
+            previousEmail, previousOrgId);
+        FinremCaseData caseData = finremCallbackRequest.getCaseDetails().getData();
+        FinremCaseData caseDataBefore = finremCallbackRequest.getCaseDetailsBefore().getData();
+
+        // Act, call performAddHearingIfNecessary using reflection
+        Method method = SolicitorAccessService.class.getDeclaredMethod(
+            "hasApplicantSolicitorChanged", FinremCaseData.class, FinremCaseData.class);
+        method.setAccessible(true);
+        boolean hasChanged = (boolean) method.invoke(solicitorAccessService, caseData, caseDataBefore);
```
[the remark this answers is not in the document]
Comment (so99y) [reply src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessServiceTest.java]: One alternative would be to create a utility class to perform hasApplicantSolicitorChanged. You can then write a test class for this new utility class.

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+        throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
+        FinremCallbackRequest finremCallbackRequest = buildCallbackRequestRespondentSolicitor(currentEmail, currentOrgId,
+            previousEmail, previousOrgId);
+        FinremCaseData caseData = finremCallbackRequest.getCaseDetails().getData();
+        FinremCaseData caseDataBefore = finremCallbackRequest.getCaseDetailsBefore().getData();
+
+        // Act, call performAddHearingIfNecessary using reflection
+        Method method = SolicitorAccessService.class.getDeclaredMethod(
+            "hasRespondentSolicitorChanged", FinremCaseData.class, FinremCaseData.class);
+        method.setAccessible(true);
```
Comment (so99y) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessServiceTest.java]: Avoiding private method testing

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AssignPartiesAccessService.java [hunk 10]
```diff
@@ -136,12 +191,22 @@ private boolean isOrgIdExists(OrganisationPolicy organisationPolicy) {
...
-            () -> log.info("{} - Attempting to grant role {} but unable to find user with email {}", caseId, caseRole, email));
+            () -> log.info("{} - Attempting to grant role {} but unable to find user", caseId, caseRole));
         if (userId.isEmpty()) {
             throw new UserNotFoundInOrganisationApiException();
         }
         return userId;
     }
+
+    private void revokeAccess(Long caseId, String email, String orgId, String caseRole)
+        throws UserNotFoundInOrganisationApiException {
```
Comment (al-hmcts) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AssignPartiesAccessService.java:218]: You don't handle null email here passed in, and also you do not need the empty check if you use `.orElseThrow`, you can do something like this. 

```java
private void revokeAccess(Long caseId, String email, String orgId, String caseRole)
        throws UserNotFoundInOrganisationApiException {
    if (StringUtils.isBlank(email)) {
        throw new UserNotFoundInOrganisationApiException(
                "{} - Attempting to revoke role {} but email prvided is blank",
                caseId,
                caseRole
        );
    }

    String userId = prdOrganisationService.findUserByEmail(email)
            .orElseThrow(() -> new UserNotFoundInOrganisationApiException(
                    "{} - Attempting to revoke role {} but system is unable to find user with provided email",
                    caseId,
                    caseRole,
            ));

    assignCaseAccessService.removeCaseRoleToUser(caseId, userId, caseRole, orgId);
}
```

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/model/ccd/FinremCaseData.java [hunk 11]
```diff
@@ -859,6 +859,20 @@ public String getRespondentSolicitorEmail() {
         return nullToEmpty(getContactDetailsWrapper().getRespondentSolicitorEmail());
     }
 
+    /**
+     * Checks if the respondent solicitor email is present and the respondent is represented by a solicitor.
+     * @return the respondent solicitor email if the respondent is represented by a solicitor, null otherwise.
+     */
+    @JsonIgnore
```
Comment (al-hmcts) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/model/ccd/FinremCaseData.java]: Looks like this method isn't used? Is it needed?

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+            .build();
+
+        FinremCaseData caseDataBefore = FinremCaseData.builder()
+            .contactDetailsWrapper(contactDetailsWrapperBefore)
+            .respondentOrganisationPolicy(OrganisationPolicy.builder().organisation(Organisation.builder()
+                .organisationID(beforeRespondentOrganisationId).build()).build())
+            .build();
+
+        return FinremCallbackRequest.builder()
+            .caseDetails(FinremCaseDetails.builder().id(123L).caseType(CONTESTED).data(caseData).build())
```
Comment (al-hmcts) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessServiceTest.java]: Can use CASE_ID_IN_LONG constant

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+            .caseDetailsBefore(FinremCaseDetails.builder().id(123L).caseType(CONTESTED).data(caseDataBefore).build())
+            .build();
+    }
+
+    private FinremCallbackRequest buildCallbackRequestRespondentSolicitor(String respondentSolicitorEmail,
+                                                                          String respondentOrganisationId,
+                                                                          String beforeRespondentSolicitorEmail,
+                                                                          String beforeRespondentOrganisationId) {
+
+        ContactDetailsWrapper contactDetailsWrapper = ContactDetailsWrapper.builder()
```
Comment (al-hmcts) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessServiceTest.java]: Could use Ashely's FinremCallbackRequestFactory here

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+                .organisationID(applicantOrganisationId).build()).build())
+            .build();
+
+        FinremCaseData caseDataBefore = FinremCaseData.builder()
+            .contactDetailsWrapper(contactDetailsWrapperBefore)
+            .applicantOrganisationPolicy(OrganisationPolicy.builder().organisation(Organisation.builder()
+                .organisationID(beforeApplicantOrganisationId).build()).build())
+            .build();
+
+        return FinremCallbackRequest.builder()
```
Comment (al-hmcts) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessServiceTest.java]: Can use FinremCallbackRequestFactory here

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+            .caseDetails(FinremCaseDetails.builder().id(123L).caseType(CONTESTED).data(caseData).build())
```
Comment (al-hmcts) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessServiceTest.java]: CASE_ID_IN_LONG

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AssignPartiesAccessService.java [hunk 16]
```diff
@@ -121,6 +120,62 @@ && isOrgIdExists(intervenerWrapper.getIntervenerOrganisation())
...
+     */
+    public void revokeApplicantSolicitor(FinremCaseData finremCaseData) throws UserNotFoundInOrganisationApiException {
+        String caseId = finremCaseData.getCcdCaseId();
+        if (finremCaseData.isApplicantRepresentedByASolicitor()
+            && isOrgIdExists(finremCaseData.getApplicantOrganisationPolicy())) {
+            String appSolicitorEmail = finremCaseData.getAppSolicitorEmailIfRepresented();
+            String appOrgId = finremCaseData.getApplicantOrganisationPolicy().getOrganisation().getOrganisationID();
+            revokeAccess(Long.valueOf(caseId), appSolicitorEmail, appOrgId, CaseRole.APP_SOLICITOR.getCcdCode());
+        } else {
+            log.info("{} - Revoke access failed - No applicant represented by a solicitor or organisation policy missing", caseId);
```
Comment (al-hmcts) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AssignPartiesAccessService.java]: throw error here? Or at least log.error

If you throw an exception then this can be handled by the the retry.executor in the submitted event then.

```
retryExecutor.runWithRetryWithHandler(
            () -> solicitorAccessService.checkAndAssignSolicitorAccess(caseData, caseDataBefore),
            "Update Contact Details - Case Solicitor Change",
            caseData.getCcdCaseId(),
            (exception, actionName, caseId) ->
                error.set(String.format("There was a problem updating solicitor access to case: %s.", exception))
        );
```

So maybe it could look something like this: 

```
public void revokeApplicantSolicitor(FinremCaseData finremCaseData) throws UserNotFoundInOrganisationApiException {
        String caseId = finremCaseData.getCcdCaseId();
        String applicantSolicitorEmail = finremCaseData.getAppSolicitorEmail();
        OrganisationPolicy applicantOrganisationPolicy = finremCaseData.getApplicantOrganisationPolicy();
        String applicantOrganisationId = applicantOrganisationPolicy.getOrganisation().getOrganisationID();

        if (!finremCaseData.isApplicantRepresentedByASolicitor() || !isOrgIdExists(applicantOrganisationPolicy)) {
            throw new IllegalStateException(
                "Unable to revoke applicant solicitor access: applicant is not represented by a solicitor or organisation policy is missing"
            );
        }

        revokeAccess(
            Long.valueOf(caseId),
            applicantSolicitorEmail,
            applicantOrganisationId,
            CaseRole.APP_SOLICITOR.getCcdCode()
        );
    }
```

And mirror for respondent

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AssignPartiesAccessService.java [hunk 17]
```diff
@@ -121,6 +120,62 @@ && isOrgIdExists(intervenerWrapper.getIntervenerOrganisation())
...
+     * log entry is recorded.
+     *
+     * @param finremCaseData the financial remedy case data before the update,
+     *                             containing previous applicant representation details and organisation information
+     */
+    public void revokeApplicantSolicitor(FinremCaseData finremCaseData) throws UserNotFoundInOrganisationApiException {
+        String caseId = finremCaseData.getCcdCaseId();
+        if (finremCaseData.isApplicantRepresentedByASolicitor()
+            && isOrgIdExists(finremCaseData.getApplicantOrganisationPolicy())) {
+            String appSolicitorEmail = finremCaseData.getAppSolicitorEmailIfRepresented();
```
Comment (al-hmcts) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AssignPartiesAccessService.java]: You don't need this method as you already check if they're represented on like 141, same goes for respondent.

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AssignPartiesAccessService.java [hunk 18]
```diff
@@ -121,6 +120,62 @@ && isOrgIdExists(intervenerWrapper.getIntervenerOrganisation())
...
+     * log entry is recorded.
+     *
+     * @param finremCaseData the financial remedy case data before the update,
+     *                             containing previous respondent representation details and organisation information
+     */
+    public void revokeRespondentSolicitor(FinremCaseData finremCaseData) throws UserNotFoundInOrganisationApiException {
+        String caseId = finremCaseData.getCcdCaseId();
+        if (finremCaseData.isRespondentRepresentedByASolicitor()
+            && isOrgIdExists(finremCaseData.getRespondentOrganisationPolicy())) {
+            String respondentSolicitorEmail = finremCaseData.getRespondentSolicitorEmailIfRepresented();
```
Comment (al-hmcts) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AssignPartiesAccessService.java]: You don't need this method as you already check if they're represented on like 169.

Comment (al-hmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AssignPartiesAccessService.java]: Just pull the sol email as you've already checked if they're repped

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/model/ccd/FinremCaseData.java (hunk 11 above)
Comment (al-hmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/model/ccd/FinremCaseData.java]: Still don't think this is needed, see comments on usage

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AssignPartiesAccessService.java (hunk 17 above)
Comment (al-hmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AssignPartiesAccessService.java]: Just pull the sol email as you've already checked if they're repped

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AssignPartiesAccessServiceTest.java [hunk 19]
```diff
@@ -287,6 +288,107 @@ void givenEmailRegistered_thenAssignAppSolicitorToCaseAndSolUidIsSet()
...
+            when(caseData.getCcdCaseId()).thenReturn(CASE_ID);
+            when(caseData.isApplicantRepresentedByASolicitor()).thenReturn(true);
+            when(caseData.getApplicantOrganisationPolicy()).thenReturn(organisationPolicy(TEST_ORG_ID));
+            when(prdOrganisationService.findUserByEmail(TEST_SOLICITOR_EMAIL)).thenReturn(Optional.empty());
+            UserNotFoundInOrganisationApiException ex = assertThrows(UserNotFoundInOrganisationApiException.class,
+                () -> assignPartiesAccessService.revokeApplicantSolicitor(caseData));
+            String message = ex.getMessage();
+            assert message != null : "Exception message should not be null";
+            assert message.contains(CASE_ID) : "Missing case id in message: " + message;
+            assert message.contains(CaseRole.APP_SOLICITOR.getCcdCode()) : "Missing case role in message: " + message;
```
Comment (so99y) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AssignPartiesAccessServiceTest.java]: why don't use assertThat

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/utils/EmailUtils.java [hunk 20]
```diff
@@ -32,4 +27,18 @@ public static boolean isValidEmailAddress(final String email, boolean allowEmpty
...
+
+    /**
+     * Determines if two solicitor email addresses are different, ignoring case. Handles nulls and empty strings as equal.
+     *
+     * @param currentEmail  the current email address
+     * @param previousEmail the previous email address
+     * @return true if the emails are different (case-insensitive), false if they are the same or both null/empty
+     */
+    public static boolean hasChange(String currentEmail, String previousEmail) {
+        if (currentEmail == null || previousEmail == null) {
```
Comment (al-hmcts) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/utils/EmailUtils.java]: Why do you still do a comparison if either arguemnt params are null? You'd want to throw an illegalAtfuement if either are null, and then do the comparision and return result.

**Review by so99y — DISMISSED** [review]

**Review by al-hmcts — DISMISSED** [review]

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsSubmittedHandler.java [hunk 21]
```diff
@@ -120,4 +140,23 @@ private void sendNocLetterToLitigants(FinremCaseDetails finremCaseDetails, Finre
...
+        FinremCaseData caseDataBefore = caseDetailsBefore.getData();
+
+        AtomicReference<String> error = new AtomicReference<>();
+
+        retryExecutor.runWithRetryWithHandler(
+            () -> solicitorAccessService.checkAndAssignSolicitorAccess(caseData, caseDataBefore),
+            "Update Contact Details - Case Solicitor Change",
+            caseData.getCcdCaseId(),
+            (exception, actionName, caseId) ->
+                error.set("There was a problem updating solicitor access to case.")
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsSubmittedHandler.java]: Adjust the message to align with the text in another PR -
https://github.com/hmcts/finrem-case-orchestration-service/pull/2913/changes#diff-f36b3c4a2c47ae6ee7fbfc267ded216911a7e5fc1b50aea76aa51d9a58bcb538R84

e.g. 
```
return "There was a problem granting access to applicant solicitor (%s). Please grant access manually."
```

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsSubmittedHandler.java [hunk 22]
```diff
@@ -77,6 +89,14 @@ public GenericAboutToStartOrSubmitCallbackResponse<FinremCaseData> handle(Finrem
         return submittedResponse();
     }
 
+    /**
+     *  Checks if the application has been issued by verifying if the issue date is present in the case details.
+     *  If the issue date is not null, it indicates that the application has been issued.
+     */
+    private boolean hasApplicationBeenIssued(FinremCaseDetails caseDetails) {
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsSubmittedHandler.java]: change the parameter to `FinremCaseData`

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsSubmittedHandler.java [hunk 23]
```diff
@@ -56,9 +61,16 @@ public GenericAboutToStartOrSubmitCallbackResponse<FinremCaseData> handle(Finrem
...
         FinremCaseData finremCaseData = caseDetails.getData();
 
-        /*
-        DFR-4589 AUto assigning applicant solicitor logic here
-        */
+        if (hasApplicationBeenIssued(caseDetails)) {
+            List<String> grantOrRevokeErrors = checkAndAssignSolicitorAccess(callbackRequest);
+
+            if (!grantOrRevokeErrors.isEmpty()) {
+                return submittedResponse(
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsSubmittedHandler.java]: Don't return here. We could adjust the notification by the grant/revoke results.

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/model/ccd/spreadsheet/FinremCaseDataTest.java
Comment (so99y) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/model/ccd/spreadsheet/FinremCaseDataTest.java:1]: I think you could remove these changes. You haven't got any changes on FinremCaseData.

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessService.java
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/SolicitorAccessService.java:1]: delete it

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsContestedMidHandler.java [hunk 24]
```diff
@@ -23,11 +24,15 @@
 public class UpdateContactDetailsContestedMidHandler extends FinremCallbackHandler {
 
     private final InternationalPostalService internationalPostalService;
+    private final ValidatePartiesService validatePartiesService;
 
     public UpdateContactDetailsContestedMidHandler(FinremCaseDetailsMapper finremCaseDetailsMapper,
-                                                   InternationalPostalService internationalPostalService) {
+                                                   InternationalPostalService internationalPostalService,
+                                                    ValidatePartiesService validatePartiesService
```
[the remark this answers is not in the document]
Comment (ptrelease) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/updatecontactdetails/UpdateContactDetailsContestedMidHandler.java]: @dawudgovuk Nope, still invalid indentation on line 31?

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/utils/EmailUtils.java
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/utils/EmailUtils.java:1]: Review is not needed. It's from https://github.com/hmcts/finrem-case-orchestration-service/pull/2927

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/utils/EmailUtilsTest.java
Comment (so99y) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/utils/EmailUtilsTest.java:1]: Review is not needed. It's from https://github.com/hmcts/finrem-case-orchestration-service/pull/2927

**Review by so99y — DISMISSED** [review]

**Review by so99y — DISMISSED** [review]

**Review by so99y — DISMISSED** [review]

**Review by so99y — APPROVED** [review]
